src/backend/common/CommandBuffer.cpp

```diff
@@ -29,6 +29,56 @@
 
 namespace backend {
 
+    namespace {
+
+        bool ValidateCopyLocationFitsInTexture(CommandBufferBuilder* builder, const TextureCopyLocation& location) {
+            const TextureBase* texture = location.texture.Get();
+            if (location.level >= texture->GetNumMipLevels()) {
+                builder->HandleError("Copy mip-level out of range");
+                return false;
+            }
+
+            // All texture dimensions are in uint32_t so by doing checks in uint64_t we avoid overflows.
+            uint64_t level = location.level;
+            if (uint64_t(location.x) + uint64_t(location.width) > (static_cast<uint64_t>(texture->GetWidth()) >> level) ||
+                uint64_t(location.y) + uint64_t(location.height) > (static_cast<uint64_t>(texture->GetHeight()) >> level)) {
+                builder->HandleError("Copy would touch outside of the texture");
+                return false;
+            }
+
+            // TODO(cwallez@chromium.org): Check the depth bound differently for 2D arrays and 3D textures
+            if (location.z != 0 || location.depth != 1) {
+                builder->HandleError("No support for z != 0 and depth != 1 for now");
+                return false;
+            }
+
+            return true;
+        }
+
+        bool FitsInBuffer(const BufferBase* buffer, uint32_t offset, uint32_t size) {
+            uint32_t bufferSize = buffer->GetSize();
+            return offset <= bufferSize && (size <= (bufferSize - offset));
+        }
+
+        bool ValidateCopySizeFitsInBuffer(CommandBufferBuilder* builder, const BufferCopyLocation& location, uint32_t dataSize) {
```

Comment:
Yes that would be amazing, and we need more compiler-stuff infrastructure for import/export symbols, defines for platforms and available backends, etc. Also this would require splitting code in common/ between common utility code (Math, compile flags, SerialQueue?, ...) and the state tracking frontend.